### ISSUE TITLE
Migrating helper/utility code out of the `core` subpackage and reorganizing

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#3314]
I would like us to move in a direction such that the `core` subpackage focuses on strictly wrapping the VTK data object types and implementing filters (with filters being nested in a sub-sub-package as they are now). We've done a great job of this so far, but a few artifacts have slipped by that seem out of place. Namely:

- `pyvista/core/common_data.py`: Why is this in the `core` module? The single method `perlin_noise` seems to me like it should be in a utils file or in the `examples` module.
- `pyvista/core/imaging.py`: Again, this is more of a utility/helper method type thing and does not pertain to the "core" API.

Further, I would like to refactor the organization of classes into more structured files following the hierarchy of the underlying VTK inheritance tree. 

![Screen Shot 2022-09-13 at 12 53 56 PM](https://user-images.githubusercontent.com/22067021/189986391-73e6124c-6dcc-49e6-ac01-80d7344a0007.png)

The file structure should generally follow this tree, for example:

- `core/`
  - `cell.py`
  - `dataset.py`: core `DataSet` wrapping
  - `datasetattributes.py`: field data wrapping
  - `errors.py`
  - `image.py`: `vtkImageData` wrapping (`pyvista.UniformGrid`)
  - `dataobject.py`: `DataObject` wrapping
  - `rectilinear.py`: `vtkRectilinearGrid` wrapping
  - `pointset/`: all `vtkPointSet` subclasses
    - `explicit.py`: `vtkExplicitStructuredGrid` wrapping
    - `points.py`: `vtkPointSet` wrapping
    - `poly.py`: `vtkPolyData` wrapping
    - `structured.py`: `vtkStructuredGrid` wrapping
    - `unstructured.py`: `vtkUnstructuredGridBase` wrapping and subclasses
  - `composite/`
    - `composite.py`: `vtkMultiBlockDataSet` wrapping
    - `partioned.py`: `vtkPartionedDataSet` wrapping

@pyvista/developers, what do you think about this reorganization? This should help make many of the files more manageable (for example, the `pyvista/core/pointset.py` is currently approaching 3,000 lines of code but will be breaking up into multiple files).

**Comment by adeak** [comment]
I see the purity argument behind following VTK's inheritence tree, but then again having a tree like
```
.
└── core
    ├── image.py that is actually UniformGrid
    ├── pointset
    │   ├── explicit.py
    │   ├── points.py for PointSet
    │   ├── poly.py
    │   ├── structured.py
    │   └── unstructured.py
    └── rectilinear.py
```
is not a bit less confusing than the current setup unless you already know how VTK stuff are structured. Uniform grids outside pointset, and called `image.py`, structured/unstructured inside pointset, rectilinear again outside... I understand that this corresponds to the diagram you showed, but it's equally confusing to me as a non-VTK-expert.

On a more bikesheddy note, I'd prefer to spell out `polydata.py` rather than just `poly.py`, and we should finally ensure that filters have a corresponding naming scheme (rather than current `dataset.py` vs `data_set.py`).

**Comment by adam-grant-hendry** [comment]
In addition, when `VTK` moved the Python modules from `vtk` to `vtkmodules` and separated classes into different submodules, it got even harder to figure out where things are. I had to implement [this script](https://stackoverflow.com/questions/73340819/where-can-i-find-information-about-the-classes-in-vtkmodules/73340820#73340820) to help me find out where classes are located:

```python
from __future__ import annotations

import importlib
import pkgutil
import warnings

import vtkmodules


def get_vtk_version() -> tuple[int, int, int]:
    """Returns the ``VTK`` version."""
    try:
        from vtkmodules.all import vtkVersion

        ver = vtkVersion()
    except AttributeError:
        warnings.warn('Unable to detect VTK version.')

    major = ver.GetVTKMajorVersion()
    minor = ver.GetVTKMinorVersion()
    patch = ver.GetVTKBuildVersion()

    return f'{major}.{minor}.{patch}'


if __name__ == '__main__':
    with open(
        f'vtkmodules_{get_vtk_version()}_hierarchy.txt',
        'w',
        encoding='utf-8',
    ) as w:
        for pkg in pkgutil.walk_packages(vtkmodules.__path__, vtkmodules.__name__ + '.'):
            try:
                module = importlib.import_module(pkg.name)
            except ImportError as err:
                continue

            w.write(f'+ {module.__name__}\n')

            for subitem in sorted(dir(module)):
                w.write(f'  - {subitem}\n')
```

[vtkmodules_9.1.0_hierarchy.txt](https://github.com/pyvista/pyvista/files/9561319/vtkmodules_9.1.0_hierarchy.txt)

This is perhaps a bit more of an aside, but it might be useful to add this to our set of utility scripts. I'm not sure how things will change from VTK version to version.

**Comment by adeak** [comment]
@adam-grant-hendry for what it's worth we at least have `pyvista.vtk_version_info`.

**Comment by adam-grant-hendry** [comment]
> @adam-grant-hendry for what it's worth we at least have `pyvista.vtk_version_info`.

Yeah, I'm pretty sure I stole the implementation of my `get_vtk_version()` from that 😉 .

> is not a bit less confusing than the current setup unless you already know how VTK stuff are structured

@adeak I tend to agree with you. Not only is there a difference between the structure of `pyvista` and `VTK`, but there is a difference in the layout of the module bindings from `VTK <8.2` and `VTK >=8.2` (see the [Python Wrappers docs](https://vtk.org/doc/nightly/html/md__builds_gitlab_kitware_sciviz_ci_Documentation_Doxygen_PythonWrappers.html)).

The layout of a static language library like VTK is very different than that of a dynamic library:

1. VTK has factory and implementation classes, so we need to distinguish between the two in hierarchies like the one you showed

2. VTK has pIMPL'd translation units, but not modules (C++ only just introduced modules into C++20). The wrappers implement namespaces as Python `module` objects, but the only namespace members that are wrapped are enum constants and types (not classes/functions or nested namespaces). The C++ namespaces could be a better indication of what the developers intend to be logically grouped together, but that's hard to see only from a graphiviz "is a" kind of relationship.

@banesullivan I think clean up is warranted, but I would caution against attempting to go too far. You will inevitably end up at the [Circle-ellipse problem](https://en.wikipedia.org/wiki/Circle%E2%80%93ellipse_problem) since it is impossible to put everything into a nice sensible hierarchy.

Things tend to work best when, given the choice between whether `A` should subclass `B` or vice-versa, the class that can reuse the most code becomes the subclass. (see [The Art of Subclassing](https://www.youtube.com/watch?v=miGolgp9xq8) starting at 29:30.)

That said, I'm all for cleaning up the structure so things are logically cleaner, but I would make the structure fit `pyvista` (not `VTK`) and leverage the strengths of Python in the process.

We're in a unique position to make the structure simpler than `VTK`. `VTK` is large and a bit stuck in its structural organization due to the language limitations of C++ and it already ships with python bindings. However the new structure of `pyvista` is chosen, I think it should also continue to serve to lower the barrier to entry to data visualization.
